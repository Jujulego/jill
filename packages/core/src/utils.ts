import { streamEvents } from '@jujulego/event-tree';
import { SpawnTask, SpawnTaskStream } from '@jujulego/tasks';

// Utils
export async function* combine<T>(...generators: AsyncGenerator<T>[]): AsyncGenerator<T> {
  for (const gen of generators) {
    yield* gen;
  }
}

export async function *streamLines(task: SpawnTask, stream: SpawnTaskStream): AsyncGenerator<string> {
  // Abort
  const ctrl = new AbortController();

  task.subscribe('status.done', () => ctrl.abort());
  task.subscribe('status.failed', () => ctrl.abort());

  // Stream
  let current = '';

  try {
    for await (const chunk of streamEvents(task, `stream.${stream}`, { signal: ctrl.signal })) {
      const data = current + chunk.data.toString('utf-8');
      const lines = data.split(/\r?\n/g);

      current = lines.pop() ?? '';

      for (const line of lines) {
        yield line;
      }
    }
  } catch (err) {
    if (task.exitCode !== 0) {
      throw err;
    }

    if (current) {
      yield current;
    }
  }
}
