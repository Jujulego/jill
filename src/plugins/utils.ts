import { type Plugin } from './types';

// Plugin utils
export function definePlugin<T, U>(plugin: Plugin<T, U>): Plugin<T, U> {
  return plugin;
}

export function assertPlugin(obj: unknown, name: string): asserts obj is Plugin {
  if (!obj) {
    throw new Error(`Plugin ${name} is not a valid plugin. Default export is null or undefined`);
  }

  if (typeof obj !== 'object') {
    throw new Error(`Plugin ${name} is not a valid plugin. Default export is a ${typeof obj}`);
  }

  if (!('builder' in obj)) {
    throw new Error(`Plugin ${name} is not a valid plugin. Missing builder method in default export`);
  }
}
