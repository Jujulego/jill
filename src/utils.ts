import { streamEvents } from '@jujulego/event-tree';
import { SpawnTask, SpawnTaskStream } from '@jujulego/tasks';
import { Arguments, Argv, CommandModule } from 'yargs';
import { Awaitable } from './types';

// Stream utils
export async function* combine<T>(...generators: AsyncGenerator<T>[]): AsyncGenerator<T> {
  for (const gen of generators) {
    yield* gen;
  }
}

export async function *streamLines(task: SpawnTask, stream: SpawnTaskStream): AsyncGenerator<string> {
  // Abort
  const ctrl = new AbortController();

  task.subscribe('completed', () => ctrl.abort());

  // Stream
  let current = '';

  try {
    for await (const chunk of streamEvents(task, `stream.${stream}`, { signal: ctrl.signal })) {
      const data = current + chunk.data.toString('utf-8');
      const lines = data.split(/\r?\n/g);

      current = lines.pop() ?? '';

      for (const line of lines) {
        yield line;
      }
    }
  } catch (err) {
    if (task.exitCode !== 0) {
      throw err;
    }

    if (current) {
      yield current;
    }
  }
}

// Command utils
export interface Middleware<T = unknown, U = unknown> {
  builder?: (yargs: Argv<T>) => Argv<U>;
  handler(args: Arguments<U>): Awaitable<void>;
}

export function applyMiddlewares<T>(yargs: Argv<T>, middlewares: Middleware[]): Argv<T> {
  let tmp: Argv<unknown> = yargs;

  for (const middleware of middlewares) {
    if (middleware.builder) {
      tmp = middleware.builder(tmp);
    }

    tmp.middleware(middleware.handler);
  }

  return tmp as Argv<T>;
}

export function defineCommand<T, U>(command: CommandModule<T, U>): CommandModule<T, U> {
  return command;
}

export function defineMiddleware<T, U>(middleware: Middleware<T, U>): Middleware<T, U> {
  return middleware;
}

// String utils
export function capitalize(txt: string): string {
  return txt.charAt(0).toUpperCase() + txt.substring(1).toLowerCase();
}
