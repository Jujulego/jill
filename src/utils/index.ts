export * from './import';
export * from './streams';
export * from './string';
export * from './yargs';
