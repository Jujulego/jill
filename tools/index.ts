export { flow } from '@jujulego/flow';
export * from './utils';
