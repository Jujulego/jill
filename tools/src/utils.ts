import { steps } from '@jujulego/flow';
import gulp from 'gulp';
import sourcemaps from 'gulp-sourcemaps';

export function src(...params: Parameters<typeof gulp.src>) {
  return steps(
    gulp.src(...params),
    sourcemaps.init()
  );
}

export function dest(path: string) {
  return steps(
    sourcemaps.write('.'),
    gulp.dest(path),
  );
}
