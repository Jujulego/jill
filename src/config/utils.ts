import { interfaces as int } from 'inversify';
import { ValidateFunction } from 'ajv';
import { cosmiconfig, defaultLoaders } from 'cosmiconfig';

import { AJV } from '@/src/ajv.config';
import schema from '@/src/assets/schema.json';
import { container } from '@/src/inversify.config';
import { dynamicImport } from '@/src/utils/import';

import { type IConfig, type IConfigExplorer } from './types';

// Symbols
export const CONFIG_EXPLORER: int.ServiceIdentifier<IConfigExplorer> = Symbol('jujulego:jill:config-explorer');
export const CONFIG_VALIDATOR: int.ServiceIdentifier<ValidateFunction<IConfig>> = Symbol('jujulego:jill:config-validator');

// Setup
container
  .bind(CONFIG_VALIDATOR)
  .toDynamicValue((context) => {
    const ajv = context.container.get(AJV);
    return ajv.compile<IConfig>(schema);
  })
  .inSingletonScope();

container
  .bind(CONFIG_EXPLORER)
  .toDynamicValue(() => {
    return cosmiconfig('jill', {
      loaders: {
        '.cjs': (filepath) => dynamicImport(filepath).then((mod) => mod.default),
        '.js': (filepath) => dynamicImport(filepath).then((mod) => mod.default),
        '.json': defaultLoaders['.json'],
        '.yaml': defaultLoaders['.yaml'],
        '.yml': defaultLoaders['.yml'],
        noExt: defaultLoaders.noExt,
      }
    });
  })
  .inSingletonScope();
