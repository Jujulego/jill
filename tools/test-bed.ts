import { type Package } from 'normalize-package-data';
import { ContainerModule } from 'inversify';
import { type CommandModule } from 'yargs';
import fs from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';

import { CONFIG } from '@/src/config/config-loader';
import { type IConfig } from '@/src/config/types';
import { container } from '@/src/inversify.config';
import { buildCommandModule, getCommandOpts, type ICommand } from '@/src/modules/command';
import { type IMiddleware } from '@/src/modules/middleware';
import { getRegistry } from '@/src/modules/module';
import { LoadProject } from '@/src/middlewares/load-project';
import { LoadWorkspace } from '@/src/middlewares/load-workspace';
import { CURRENT } from '@/src/constants';
import { type Project } from '@/src/project/project';
import { Workspace } from '@/src/project/workspace';
import { type PackageManager } from '@/src/project/types';
import { type Class } from '@/src/types';

import { TestProject } from './test-project';
import { TestWorkspace } from './test-workspace';
import { shell } from './utils';
import { ContextService } from '@/src/commons/context.service';

// Bed
export class TestBed {
  // Attributes
  private _config: IConfig = {};

  readonly project = new TestProject('./test');

  // Methods
  manifest(pkg: Partial<Package>): Package {
    return {
      _id: 'test-id',
      name: 'test',
      version: '1.0.0',
      readme: 'readme',
      ...pkg
    };
  }

  addWorkspace(name: string, pkg: Partial<Package> = {}): TestWorkspace {
    const wks = new TestWorkspace(`./${name}`, this.manifest({ _id: `${name}-id`, name, ...pkg }), this.project);
    this.project.addWorkspace(wks);

    return wks;
  }

  async writeManifest(path: string, wks: Workspace): Promise<void> {
    const { _id: _, ...manifest } = wks.manifest;
    await fs.writeFile(path, JSON.stringify(manifest));
  }

  /**
   * Loads command, and mock LoadProject & LoadWorkspace middlewares
   *
   * @param command Command to prepare
   * @param within project or workspace where the comme will be run
   */
  async prepareCommand(command: Class<ICommand>, within: Project | Workspace = this.project): Promise<CommandModule> {
    // Load metadata
    const opts = getCommandOpts(command);
    const registry = getRegistry(command);

    // Create command
    container.load(new ContainerModule(registry));
    const cmd = await container.getAsync<ICommand>(command);

    // Inject mocks
    const prj = within instanceof Workspace ? within.project : within;
    const wks = within instanceof Workspace ? within : await within.mainWorkspace();

    container.rebind<IMiddleware>(LoadProject).toConstantValue({
      handler() {
        container.get(ContextService).project = prj;
      }
    });

    container.rebind<IMiddleware>(LoadWorkspace).toConstantValue({
      handler() {
        container.get(ContextService).workspace = wks;
      }
    });

    return buildCommandModule(cmd, opts);
  }

  /**
   * Create project's structure inside a temporary directory
   */
  async createProjectDirectory(): Promise<string> {
    // Ensure tmp dir exists (for mocked fs)
    await fs.mkdir(os.tmpdir(), { recursive: true });

    let tmp = await fs.mkdtemp(path.join(os.tmpdir(), 'jill-test-'));

    // Corrects path on macOS => see https://github.com/nodejs/node/issues/11422
    tmp = await fs.realpath(tmp);

    // Create project directory
    const prjDir = path.join(tmp, 'test');

    await fs.mkdir(prjDir);
    await this.writeManifest(path.join(prjDir, 'package.json'), await this.project.mainWorkspace());

    // Add config file
    await fs.writeFile(path.join(prjDir, '.jillrc.json'), JSON.stringify(this._config));

    // Create workspaces
    for await (const wks of this.project.workspaces()) {
      const wksDir = path.join(prjDir, wks.name);

      await fs.mkdir(wksDir);
      await this.writeManifest(path.join(wksDir, 'package.json'), wks);
    }

    return prjDir;
  }

  /**
   * Create project's structure inside a temporary directory and generates the lock file
   */
  async createProjectPackage(pm: PackageManager): Promise<string> {
    const prjDir = await this.createProjectDirectory();

    // Run package manager
    switch (pm) {
      case 'npm':
        await shell('npm install', { cwd: prjDir });
        break;

      case 'yarn':
        await shell('yarn install --no-immutable', { cwd: prjDir });
        break;
    }

    return prjDir;
  }

  // Properties
  get config(): Readonly<IConfig> {
    return this._config;
  }

  set config(config) {
    this._config = config;

    container.rebind(CONFIG).toConstantValue(config);
  }
}
