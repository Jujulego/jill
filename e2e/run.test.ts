import fs from 'node:fs/promises';
import path from 'node:path';

import { TestBed } from '@/tools/test-bed';
import { fileExists, noColor } from '@/tools/utils';

import { usePackageManager, jill } from './utils';

describe('jill run', () => void usePackageManager((packageManager) => {
  // Setup
  let prjDir: string;

  beforeEach(async () => {
    const bed = new TestBed();

    const wksC = bed.addWorkspace('wks-c', {
      scripts: {
        // language=bash
        build: 'node -e "require(\'node:fs\').writeFileSync(\'script.txt\', \'build\')"',
        // language=bash
        start: 'node -e "require(\'node:fs\').writeFileSync(\'script.txt\', \'start\')"',
        // language=bash
        fails: 'node -e "process.exit(1)"',
      }
    });
    const wksB = bed.addWorkspace('wks-b', {
      scripts: {
        // language=bash
        start: 'node -e "require(\'node:fs\').writeFileSync(\'script.txt\', \'start\')"'
      }
    })
      .addDependency(wksC, true);
    bed.addWorkspace('wks-a')
      .addDependency(wksB)
      .addDependency(wksC, true);

    prjDir = await bed.createProjectPackage(packageManager);
  });

  afterEach(async () => {
    await fs.rm(prjDir, { recursive: true });
  });

  // Tests
  it('should run wks-c start script', async () => {
    const res = await jill(['run', '-w', 'wks-c', 'start'], { cwd: prjDir });

    // Check jill output
    expect(res.code).toBe(0);

    const screen = res.screen.screen.split('\n')
      .filter((line) => !noColor(line).startsWith('[wks-c#start]'));

    expect(screen).toMatchLines([
      expect.ignoreColor(/. Running start in wks-c \(took [0-9.]+m?s\)/),
    ]);

    // Check script result
    await expect(fs.readFile(path.join(prjDir, 'wks-c', 'script.txt'), 'utf8'))
      .resolves.toBe('start');
  });

  it('should run wks-c fails script and exit 1', async () => {
    const res = await jill(['run', '-w', 'wks-c', 'fails'], { cwd: prjDir });

    // Check jill output
    expect(res.code).toBe(1);

    const screen = res.screen.screen.split('\n')
      .filter((line) => !noColor(line).startsWith('[wks-c#fails]'));

    expect(screen).toMatchLines([
      expect.ignoreColor(/. Running fails in wks-c \(took [0-9.]+m?s\)/),
    ]);
  });

  it('should run wks-b start script and build script', async () => {
    const res = await jill(['run', '-w', 'wks-b', 'start'], { cwd: prjDir });

    // Check jill output
    expect(res.code).toBe(0);

    const screen = res.screen.screen.split('\n')
      .filter((line) => !noColor(line).startsWith('[wks-c#build]'))
      .filter((line) => !noColor(line).startsWith('[wks-b#start]'));

    expect(screen).toMatchLines([
      expect.ignoreColor(/. Running build in wks-c \(took [0-9.]+m?s\)/),
      expect.ignoreColor(/. Running start in wks-b \(took [0-9.]+m?s\)/),
    ]);

    // Check scripts result
    await expect(fs.readFile(path.join(prjDir, 'wks-c', 'script.txt'), 'utf8'))
      .resolves.toBe('build');

    await expect(fs.readFile(path.join(prjDir, 'wks-b', 'script.txt'), 'utf8'))
      .resolves.toBe('start');
  });

  it('should print task plan and do not run any script', async () => {
    const res = await jill(['run', '-w', 'wks-b', '--plan', '--planMode', 'json', 'start'], { cwd: prjDir });

    // Check jill plan
    expect(res.code).toBe(0);
    expect(res.stdout).toHaveLength(1);

    const plan = JSON.parse(res.stdout[0]);
    expect(plan).toHaveLength(2);

    expect(plan[0]).toMatchSnapshot({
      id: expect.stringMatching(/[0-9a-f]{32}/),
      context: {
        workspace: {
          cwd: expect.any(String)
        }
      }
    });
    expect(plan[0].context.workspace.cwd).toBe(path.join(prjDir, 'wks-c'));

    expect(plan[1]).toMatchSnapshot({
      id: expect.stringMatching(/[0-9a-f]{32}/),
      dependenciesIds: [
        expect.stringMatching(/[0-9a-f]{32}/)
      ],
      context: {
        workspace: {
          cwd: expect.any(String)
        }
      }
    });
    expect(plan[1].dependenciesIds).toEqual([plan[0].id]);
    expect(plan[1].context.workspace.cwd).toBe(path.join(prjDir, 'wks-b'));

    await expect(fileExists(path.join(prjDir, 'wks-c', 'script.txt'))).resolves.toBe(false);
    await expect(fileExists(path.join(prjDir, 'wks-b', 'script.txt'))).resolves.toBe(false);
  });
}));
