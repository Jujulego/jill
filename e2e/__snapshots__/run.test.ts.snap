// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jill run using npm should print task plan and do not run any script 1`] = `
"[1mId[22m      [1mName[22m           [1mWorkspace[22m  [1mGroup[22m  [1mDepends on[22m
4bbdcd  npm run build  wks-c
1f291e  npm run start  wks-b             4bbdcd"
`;

exports[`jill run using yarn should print task plan and do not run any script 1`] = `
"[1mId[22m      [1mName[22m            [1mWorkspace[22m  [1mGroup[22m  [1mDepends on[22m
8f3501  yarn run build  wks-c
4a3158  yarn run start  wks-b             8f3501"
`;
